keps/sig-api-machinery/365-paginated-lists/README.MD

```diff
@@ -152,6 +248,86 @@ malicious users cannot trick the server into serving results outside of its rang
 The storage layer must perform defensive checking on the provided value, check for path traversal attacks, and have 
 stable versioning for the continue token.
 
+#### watchcache
+
+However, not all LIST requests are served from etcd. Our API allows clients to opt-in for the LIST result to be
+served from kube-apiserver cache (called watchcache), being propagated via watch from etcd3.
+However, up until release 1.24, support for pagination wasn't added to watchcache. This makes the logic that
+we use extremely non-intuitive, as it effectively works as following:
+
+- if the `continuation` is set - always pass the request to etcd (and honor the LIMIT)
+- if resourveVersion='0' - ignore the LIMIT parameter and return result from watchcache
+- if resourceVersion=<some concrete RV> and LIMIT is set - pass the request to etcd (and honor the LIMIT)
+- if resourceVersion=<some concrete RV> and LIMIT is not set - serve the request from watchcache
+   - which effectively results in different semantics about data freshness depending on LIMIT being set or not
+
+Fixing that misleading experience is one of the issues that are GA blocker for the whole pagination feature.
+
+We propose to add support for pagination in watchcache as following:
+
+1. As of now, watchache (in addition to transaction log which is not relevant for this proposal) contains
+  the current (consistent at a given RV) state of all objects of a given type. This is currently stored
+  in client-go `Indexer`, which internally stores that in hash-map.
+  We propose to replace the usage of `Indexer` with an implementation that will be build on top of
+  BTree that supports copy-on-write semantic. We are planning to use the implementation from
+  [https://github.com/google/btree] (although it may require adding more extensive test coverage).
+1. When a request with LIMIT set reaches watchcache, we clone (create a copy-onwrite copy) of the current
```

Comment:
I am not an expert but this sounds like what `LMDB/bbolt` does, could be worth exploring if we decide to go this path.